rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -158,22 +194,58 @@ private void parseIfNotParsing() {
             if (parsing.compareAndSet(false, true)) {
                 parser
                     .apply(location, content)
-                    .thenApply(t -> new Versioned<>(version, t))
-                    .whenComplete((t, error) -> {
-                        if (t != null) {
-                            var errors = RECOVERY.findAllErrors(t.get());
-                            if (errors.isEmpty()) {
-                                Versioned.replaceIfNewer(lastWithoutErrors, t);
+                    .whenComplete((t, e) -> {
+
+                        // Prepare result values for futures
+                        var tree = new Versioned<>(version, t);
+                        var diagnostics = new Versioned<>(version, toDiagnostics(t, e));
+
+                        // Complete future to get the tree
+                        if (t == null) {
+                            treeAsync.completeExceptionally(e);
+                        } else {
+                            treeAsync.complete(tree);
+                            Versioned.replaceIfNewer(last, tree);
+                            if (diagnostics.get().isEmpty()) {
+                                Versioned.replaceIfNewer(lastWithoutErrors, tree);
                             }
-                            Versioned.replaceIfNewer(last, t);
-                            treeAsync.complete(t);
-                        }
-                        if (error != null) {
-                            treeAsync.completeExceptionally(error);
                         }
+
+                        // Complete future to get diagnostics
+                        diagnosticsAsync.complete(diagnostics);
                     });
             }
         }
+
+        private List<Diagnostic> toDiagnostics(ITree tree, Throwable excp) {
```

Comment:
The code in this method is copied verbatim from `RascalTextDocumentService`. I might like to refactor it a little bit after the first review round, but thought it'd be easier to keep it like this for now.